src/lib.rs

```diff
@@ -56,6 +56,103 @@ fn default_build_command() -> process::Command {
     command
 }
 
+fn install_wasm_opt(target_path: impl AsRef<Path>) -> Result<PathBuf> {
+    let cache = binary_install::Cache::at(target_path.as_ref());
+    let url = format!(
+        "https://github.com/WebAssembly/binaryen/releases/download/version_{version}/binaryen-version_{version}-{arch}-{os}.tar.gz",
+        version = "105",
+        arch = platforms::TARGET_ARCH,
+        os = platforms::TARGET_OS,
+    );
+
+    #[cfg(target_os = "macos")]
+    let binaries = &["wasm-opt", "libbinaryen"];
+    #[cfg(not(target_os = "macos"))]
+    let binaries = &["wasm-opt"];
+
+    log::info!("Downloading wasm-opt");
+    Ok(cache
+        .download(true, "wasm-opt", binaries, &url)
+        .map_err(|err| err.compat())
+        .with_context(|| format!("could not download binaryen: {}", url))?
+        .expect("cannot install binaryen")
+        .binary("wasm-opt")
+        .map_err(|err| err.compat())?)
+}
+
+#[allow(unused_variables, unreachable_code)]
+pub fn wasm_opt(
+    binary: Vec<u8>,
+    shrink_level: u32,
+    optimization_level: u32,
+    debug_info: bool,
+    target_path: impl AsRef<Path>,
```

Comment:
You don't need the user to give you the target directory. You can find it on your own